mailbox/api/src/main/java/org/apache/james/mailbox/model/SearchQuery.java

```diff
@@ -759,49 +760,82 @@ public static Criterion mimeMessageID(String messageId) {
         return new MimeMessageIDCriterion(messageId);
     }
 
-    private final Set<MessageUid> recentMessageUids = new HashSet<>();
+    public static class Builder {
+        private final ImmutableList.Builder<Criterion> criterias;
+        private final ImmutableSet.Builder<MessageUid> recentMessageUids;
+        private Optional<ImmutableList<Sort>> sorts;
 
-    private final List<Criterion> criterias;
+        public Builder() {
+            criterias = ImmutableList.builder();
+            sorts = Optional.empty();
+            recentMessageUids = ImmutableSet.builder();
+        }
 
-    private List<Sort> sorts = Collections.singletonList(new Sort(Sort.SortClause.Uid, Sort.Order.NATURAL));
+        public Builder andCriteria(Criterion... criteria) {
+            return andCriteria(Arrays.asList(criteria));
+        }
 
-    public SearchQuery(Criterion... criterias) {
-        this(new ArrayList<>(Arrays.asList(criterias)));
-    }
+        public Builder andCriteria(Collection<Criterion> criteria) {
+            this.criterias.addAll(criteria);
+            return this;
+        }
+
+        public Builder sorts(Sort... sorts) {
+            return this.sorts(Arrays.asList(sorts));
+        }
 
-    public SearchQuery() {
-        this(new ArrayList<>());
+        public Builder sorts(List<Sort> sorts) {
+            if (sorts == null || sorts.isEmpty()) {
+                throw new IllegalArgumentException("There must be at least one Sort");
+            }
+            this.sorts = Optional.of(ImmutableList.copyOf(sorts));
+            return this;
+        }
+
+        public Builder addRecentMessageUids(Collection<MessageUid> uids) {
+            recentMessageUids.addAll(uids);
+            return this;
+        }
+
+        public SearchQuery build() {
+            return new SearchQuery(criterias.build(),
+                sorts.orElse(ImmutableList.of(new Sort(Sort.SortClause.Uid, Sort.Order.NATURAL))),
+                recentMessageUids.build());
+        }
     }
 
-    private SearchQuery(List<Criterion> criterias) {
-        this.criterias = criterias;
+    public static Builder builder() {
+        return new Builder();
     }
 
-    public void andCriteria(Criterion crit) {
-        criterias.add(crit);
+    public static SearchQuery of(Criterion... criterias) {
```

Comment:
what meaning has a list of zero criteria?

Reply:
It means 'matchAll'. I propose to add a new method to handle it.